### ISSUE TITLE
Fasta stack trace from BodyBuilder

### ISSUE THREAD

**Issue by lukechurch** [dart-lang/sdk#29937]
@peter-ahe-google 
ID: 08ffa49d09f3b3c7c0b3c86c45c479b89ebbbd78
```
Unhandled exception:
Unhandled: FunctionNode
#0      internalError (package:front_end/src/fasta/errors.dart:38:5)
#1      BodyBuilder.toValue (package:front_end/src/fasta/kernel/body_builder.dart:224:14)
#2      BodyBuilder.popForValue (package:front_end/src/fasta/kernel/body_builder.dart:205:31)
#3      BodyBuilder.popListForValue (package:front_end/src/fasta/kernel/body_builder.dart:237:17)
#4      BodyBuilder.handleLiteralList (package:front_end/src/fasta/kernel/body_builder.dart:1538:36)
#5      Parser.parseLiteralListSuffix (package:front_end/src/fasta/parser/parser.dart:3115:16)
#6      Parser.parsePrimary (package:front_end/src/fasta/parser/parser.dart:3011:14)
#7      Parser.parseUnaryExpression (package:front_end/src/fasta/parser/parser.dart:2942:14)
#8      Parser.parsePrecedenceExpression (package:front_end/src/fasta/parser/parser.dart:2811:13)
#9      Parser.parseExpression (package:front_end/src/fasta/parser/parser.dart:2784:11)
#10     Parser.parseArguments (package:front_end/src/fasta/parser/parser.dart:3441:15)
#11     Parser.parseArgumentsOpt (package:front_end/src/fasta/parser/parser.dart:3415:14)
#12     Parser.parseSend (package:front_end/src/fasta/parser/parser.dart:3395:13)
#13     Parser.parseSendOrFunctionLiteral (package:front_end/src/fasta/parser/parser.dart:3219:14)
#14     Parser.parsePrimary (package:front_end/src/fasta/parser/parser.dart:2972:14)
#15     Parser.parseUnaryExpression (package:front_end/src/fasta/parser/parser.dart:2942:14)
#16     Parser.parsePrecedenceExpression (package:front_end/src/fasta/parser/parser.dart:2811:13)
#17     Parser.parseExpression (package:front_end/src/fasta/parser/parser.dart:2784:11)
#18     Parser.parseExpressionStatement (package:front_end/src/fasta/parser/parser.dart:2700:13)
#19     Parser.parseExpressionStatementOrDeclaration (package:front_end/src/fasta/parser/parser.dart:2651:12)
#20     Parser.parseStatementX (package:front_end/src/fasta/parser/parser.dart:2441:14)
#21     Parser.parseStatement (package:front_end/src/fasta/parser/parser.dart:2433:20)
#22     Parser.parseFunctionBody (package:front_end/src/fasta/parser/parser.dart:2363:15)
#23     DietListener.parseFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:545:14)
#24     DietListener.buildFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:433:5)
#25     DietListener.endTopLevelMethod (package:front_end/src/fasta/source/diet_listener.dart:214:5)
#26     Parser.parseTopLevelMethod (package:front_end/src/fasta/parser/parser.dart:1427:14)
#27     Parser.parseTopLevelMember (package:front_end/src/fasta/parser/parser.dart:1333:11)
#28     Parser._parseTopLevelDeclaration (package:front_end/src/fasta/parser/parser.dart:372:14)
#29     Parser.parseTopLevelDeclaration (package:front_end/src/fasta/parser/parser.dart:344:13)
#30     Parser.parseUnit (package:front_end/src/fasta/parser/parser.dart:336:15)
#31     SourceLoader.buildBody (package:front_end/src/fasta/source/source_loader.dart:146:14)
<asynchronous suspension>
#32     Loader.buildBodies (package:front_end/src/fasta/loader.dart:117:13)
<asynchronous suspension>
#33     KernelTarget.buildProgram (package:front_end/src/fasta/kernel/kernel_target.dart:282:20)
<asynchronous suspension>
#34     CompileTask.compile (package:front_end/src/fasta/fasta.dart:164:38)
<asynchronous suspension>
#35     compile.<anonymous closure> (package:front_end/src/fasta/fasta.dart:101:25)
<asynchronous suspension>
#36     CompilerContext.withGlobalOptions.<anonymous closure> (package:front_end/src/fasta/compiler_context.dart:45:33)
#37     _rootRun (dart:async/zone.dart:1120)
#38     _CustomZone.run (dart:async/zone.dart:1001)
#39     runZoned (dart:async/zone.dart:1467)
#40     CompilerContext.withGlobalOptions (package:front_end/src/fasta/compiler_context.dart:45:12)
#41     CompilerCommandLine.withGlobalOptions (package:front_end/src/fasta/compiler_command_line.dart:120:28)
#42     compile (package:front_end/src/fasta/fasta.dart:94:38)
<asynchronous suspension>
#43     compileEntryPoint (package:front_end/src/fasta/fasta.dart:53:11)
<asynchronous suspension>
#44     main (file:///home/lukechurch/GitRepos/sdk/sdk/pkg/front_end/tool/_fasta/compile.dart:7:33)
#45     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:263)
#46     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

**Comment by peter-ahe-google** [comment]
@lukechurch ID missing.

**Comment by lukechurch** [comment]
@peter-ahe-google PTAL

**Comment by peter-ahe-google** [comment]
Repro:

```
main() {
  [f() {}];
}
```

Almost identical to #29979.

**Comment by peter-ahe-google** [comment]
Proposed fix in [CL 2968093003](https://codereview.chromium.org/2968093003).
